drud/ddev
# Always set DDEV_UID/DDEV_GUID to run web container as host uid, fixes #635, fixes #740

## The Problem/Issue/Bug:

We need to run our web and db containers predictably as the same uid and gid as the host user. This also prevents security issues where a container process might create files with permissions we don't want.

#635 (File write permissions fail on linux) is one example of this general problem.
#740 (web container fails on gid collision) is another example

## How this PR Solves The Problem:

The web and db containers are reworked to be able to run as an unprivileged user. This situation resolves all the problems associated with running the container as root. However, it doesn't work everywhere. Linux containers can have UIDs only up to 60000, and the "uid" equivalent on Windows may be much higher than that. 

As a result, on Windows or whenever the UID is high, the containers are still run as root. This is logic decided in ddev.

## Manual Testing Instructions:

* Remove .ddev from the project
* ddev config
* ddev start
* Use the site as normally intended. You can `ddev ssh` to see the uid that the user is running as, `ps -ef` in there to see what user the important processes are running as.
* Try some composer commands in there. 

## Automated Testing Overview:

The same automated testing is in play, as no functionality really changes, except the warning to the user when containers are running as root.

## Related Issue Link(s):

#635 
#740 
web container PR: https://github.com/drud/docker.nginx-php-fpm-local/pull/60
db container PR: https://github.com/drud/mariadb-local/pull/51

## Release/Deployment notes:


- [ ] Make new release and push for docker.nginx-php-fpm-local
- [x] Make new release and push for mariadb-local
- [ ] Update this PR with the new released tags

## Security Implications

In the vast majority of cases on Linux or Mac, the web and db containers will run as an unprivileged user. This solves all kinds of file sharing and access problems.

In another set of cases (Docker for Windows and Docker for Mac) the Docker environment prevents any bad coming from containers even if they are run with root privileges (as seems necessary on Windows).

In the remaining cases (say a user ID > 60000 on linux, or some Windows situation we don't have experience with, not Docker for Windows) the container could be run as root and have privileged access to the filesystem where it's being run. We don't have any knowledge of this situation or how it can occur, but it probably would not be best from a security perspective.
